src/apsbits/core/instrument_init.py

```diff
@@ -143,77 +165,35 @@ def namespace_loader(yaml_device_file, main=True):
-        if isinstance(config_file, str):
-            config_file = pathlib.Path(config_file)
 
-        def parser(creator, specs):
-            if creator not in self.device_classes:
-                try:
-                    self.device_classes[creator] = dynamic_import(creator)
-                except ImportError as e:
-                    logger.error(
-                        "Failed to import device creator '%s': %s", creator, str(e)
-                    )
-                    raise
-                except AttributeError as e:
-                    logger.error(
-                        "Device creator '%s' not found in module: %s", creator, str(e)
-                    )
-                    raise
-            entries = [
-                {
-                    "device_class": creator,
-                    "args": (),  # ALL specs are kwargs!
-                    "kwargs": table,
-                }
-                for table in specs
-            ]
-            return entries
-
-        try:
-            with open(config_file, "r") as f:
-                config_data = load_config_yaml(f)
-        except FileNotFoundError:
-            logger.error("Device configuration file not found: %s", config_file)
-            raise
-        except PermissionError:
-            logger.error("Permission denied reading device file: %s", config_file)
-            raise
-        except yaml.YAMLError as e:
-            logger.error(
-                "YAML parsing error in device file %s: %s", config_file, str(e)
-            )
-            raise
-
-        if not isinstance(config_data, dict):
-            logger.error(
-                "Invalid device file format in %s: expected dictionary, got %s",
-                config_file,
-                type(config_data).__name__,
-            )
-            raise ValueError(f"Invalid device file format in {config_file}")
-
-        try:
-            devices = [
-                device
-                # parse the file using already loaded config data
-                for k, v in config_data.items()
-                # each support type (class, factory, function, ...)
-                for device in parser(k, v)
-            ]
-        except Exception as e:
-            logger.error(
-                "Error parsing device specifications in %s: %s", config_file, str(e)
-            )
-            raise
-        return devices
-
-
-instrument = Instrument({})  # singleton
-oregistry = instrument.devices
-"""Registry of all ophyd-style Devices and Signals."""
+def with_registry(func: Callable) -> Callable:
+    """
+    Decorator that provides access to the instrument's device registry.
+    Injects 'oregistry' as the first argument to the decorated function.
+    """
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        if _instrument is None:
+            raise RuntimeError("Instrument not set. Call set_instrument() first.")
+        return func(_instrument.devices, *args, **kwargs)
```

Comment:
Grabbing the first positional argument is a bit surprising and does not give the user anyway to over-ride the behavior.  I suggest doing

```suggestion
    def wrapper(*args, **kwargs, oregistry=None):
        if oregistry is None:
            if _instrument is None:
                raise RuntimeError("Instrument not set. Call set_instrument() first.")
             else:
                 oregistry = _instrument.devices
        if 
        return func(*args, **kwargs, oregistry=oregistry)
```

and specify it as a required keykword argument in the wrapped functions like

```python

@with_registry
def plan(..., *, oregistry):
    ...

```